### PR TITLE
Handle constant B-spline mutual information inputs

### LINKED ISSUE #9018: GlobalMutualInformationLoss returns NaN for constant B-spline inputs
**Describe the bug**

`GlobalMutualInformationLoss(kernel_type="b-spline")` returns a non-finite loss for valid constant prediction or target tensors. Constant tensors have zero intensity range, so the B-spline bin width becomes zero and normalization produces NaNs.

Two related boundary failures exist in the same path:

- `num_bins <= 4` cannot accommodate the implementation's two-bin padding on each side, leaving a zero or negative interior denominator.
- In float16, the cubic B-spline distance polynomial can overflow before masking.

**To Reproduce**

```python
import torch
from monai.losses import GlobalMutualInformationLoss

pred = torch.zeros((1, 1, 8, 8), requires_grad=True)
target = torch.ones_like(pred)
loss = GlobalMutualInformationLoss(kernel_type="b-spline")
result = loss(pred, target)
print(result, torch.isfinite(result))
```

On the current `dev` revision, `result` is non-finite.

**Expected behavior**

Valid constant tensors should produce a finite loss and finite gradients. B-spline configurations with too few bins should fail at construction with a clear `ValueError`.

**Screenshots**

Not applicable.

**Environment**

```text
MONAI version: 1.6.0rc1+46.g3ee058bd
MONAI revision: 3ee058bdd16dd4a566d23d3f84687c3c35268a36
Python version: 3.13.13
PyTorch version: 2.12.0
NumPy version: 2.4.6
System: macOS 27.0 arm64
CUDA available: False
```

**Additional context**

A focused fix and regression coverage are proposed in #9022.

### PR DESCRIPTION
**Pull request by yinkev**
Fixes #9018.

### Description

B-spline global mutual information now remains finite when either prediction or target has a constant intensity range and for float16, bfloat16, or float32 inputs with tiny nonzero ranges. B-spline configurations with four or fewer bins are rejected because the implementation reserves two padding bins at each boundary. Gaussian configurations with one or fewer bins are rejected before bin-centre spacing is computed.

The fix performs reduced-precision range, bin-width, and B-spline coordinate arithmetic in float32. Bin widths too small to be represented safely by the compute or input dtype are treated as a degenerate range. This prevents zero-range division, tiny-range underflow, unrepresentable normalization slopes, and cubic-distance overflow while preserving autograd back to the original input.

### Types of changes

- [x] A Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [x] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### Validation

- zero-range and tiny-range regressions fail before the follow-up fix;
- prediction-constant and target-constant paths are covered independently;
- extreme-range coverage includes float16, bfloat16, and float32 tiny ranges plus a large float16 range with `num_bins=64`;
- a one-bin Gaussian constructor regression verifies invalid spacing is rejected;
- exact final head `354e5975` affected loss module: 24 passed, 1 skipped;
- ordinary float32/float64 outputs and gradients remain unchanged in differential checks;
- Black, isort, Ruff, DCO, and `git diff --check` pass.

### CODE REVIEW

File: monai/losses/image_dissimilarity.py [hunk 1]
```diff
@@ -224,13 +224,20 @@ def __init__(
...
         """
         super().__init__(reduction=LossReduction(reduction).value)
         if num_bins <= 0:
             raise ValueError(f"num_bins must > 0, got {num_bins}")
         bin_centers = torch.linspace(0.0, 1.0, num_bins)  # (num_bins,)
         sigma = torch.mean(bin_centers[1:] - bin_centers[:-1]) * sigma_ratio
         self.kernel_type = look_up_option(kernel_type, ["gaussian", "b-spline"])
+        # B-spline windowing reserves two padding bins at each boundary.
+        if self.kernel_type == "b-spline" and num_bins <= 4:
+            raise ValueError(f"num_bins must be greater than 4 for b-spline kernel, got {num_bins}")
```
[the remark this answers is not in the document]
Comment (yinkev) [reply monai/losses/image_dissimilarity.py:241]: Verified, but intentionally not bundled. Gaussian `num_bins=1` is a pre-existing configuration defect independent of the B-spline zero/tiny-range bug tracked by #9018; the PR body records it as separate follow-up scope.

File: monai/losses/image_dissimilarity.py [hunk 2]
```diff
@@ -283,7 +290,9 @@ def parzen_windowing_b_spline(self, img: torch.Tensor, order: int) -> tuple[torc
         # window.
         _max, _min = torch.max(img), torch.min(img)
         padding = 2
-        bin_size = (_max - _min) / (self.num_bins - 2 * padding)
+        value_range = _max - _min
+        bin_size = value_range / (self.num_bins - 2 * padding)
+        bin_size = torch.where(value_range > 0, bin_size, torch.ones_like(bin_size))
```
[the remark this answers is not in the document]
Comment (yinkev) [reply monai/losses/image_dissimilarity.py]: Fixed in db5f54a9. Reduced-precision min/max, range, bin-width, and coordinate arithmetic now stay in float32, with zero/tiny/large-range regressions. The safe bin width is deliberately not cast back to half because that would recreate underflow.

File: tests/losses/image_dissimilarity/test_global_mutual_information_loss.py [hunk 3]
```diff
@@ -164,6 +169,38 @@ def test_ill_opts(self, num_bins, reduction, expected_exception, expected_messag
...
 
 
+class TestGlobalMutualInformationLossBSpline(unittest.TestCase):
+    """Test B-spline mutual information on degenerate intensity ranges."""
+
+    def test_b_spline_constant_images_are_finite(self):
+        """Verify constant float32 inputs yield finite loss and gradients."""
+        pred = torch.zeros((1, 1, 8, 8), requires_grad=True)
+        target = torch.ones_like(pred)
+        loss = GlobalMutualInformationLoss(kernel_type="b-spline")
```
[the remark this answers is not in the document]
Comment (yinkev) [reply tests/losses/image_dissimilarity/test_global_mutual_information_loss.py]: Fixed in db5f54a9. Constant prediction and constant target are exercised independently, with finite loss and gradient assertions.

File: tests/losses/image_dissimilarity/test_global_mutual_information_loss.py [hunk 4]
```diff
@@ -164,6 +169,38 @@ def test_ill_opts(self, num_bins, reduction, expected_exception, expected_messag
...
+        self.assertAlmostEqual(result.item(), 0.0, places=6)
+        result.backward()
+        self.assertIsNotNone(pred.grad)
+        self.assertTrue(torch.isfinite(pred.grad).all())
+
+    def test_b_spline_constant_half_precision_images_are_finite(self):
+        """Verify constant float16 inputs yield finite loss and gradients."""
+        pred = torch.zeros((1, 1, 8, 8), dtype=torch.float16, requires_grad=True)
+        target = torch.ones_like(pred)
+        loss = GlobalMutualInformationLoss(kernel_type="b-spline", num_bins=32)
```
[the remark this answers is not in the document]
Comment (yinkev) [reply tests/losses/image_dissimilarity/test_global_mutual_information_loss.py]: Fixed in db5f54a9. The half-precision B-spline regression uses 64 bins so the cubic distance path exceeds float16 range before the float32 compute fix.

**Comment by yinkev** [comment]
CI triage: the only failed check is `full-dep (macOS-latest)`. The affected `GlobalMutualInformationLoss` tests passed in that job; the failure occurred later in `tests.data.test_persistentdataset_dist.TestDistDataset` because Gloo could not resolve the ephemeral runner hostname. Ubuntu and Windows full-dependency jobs and all other checks passed. This is unrelated to the PR diff. Fork contributors cannot rerun upstream Actions, so the matrix will run again when this draft is rebased before promotion.

**Comment by yinkev** [comment]
The only failing check is the macOS full-dep job (89732744347), in the unrelated distributed PersistentDataset test: Gloo cannot resolve the runner .local hostname (`Unable to find address for ...local`). Linux, Windows, static, packaging, and docs checks passed. I attempted to rerun only the failed job, but GitHub requires repository admin permission for the rerun. A maintainer rerun should be sufficient; no PR code change appears warranted.

**Comment by yinkev** [comment]
Addressed the latest CodeRabbit documentation nit in `de7cbc17`: the parameterized B-spline regressions now document `constant_input`, the case label, `dtype`, and `maximum` with Google-style `Args` sections. The six affected cases and the complete 21-test file pass; Ruff check/format and `git diff --check` pass.

File: monai/losses/image_dissimilarity.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (yinkev) [reply monai/losses/image_dissimilarity.py:241]: Fixed in 16047535. Gaussian configurations now reject `num_bins <= 1` before bin-centre spacing is computed, while the existing B-spline `num_bins <= 4` constraint remains unchanged. A public-constructor regression covers the one-bin Gaussian case. The exact regression passed, the full affected module passed (23 passed, 1 skipped), and the MONAI copyright/isort/Black/Ruff gate plus `git diff --check` passed.

**Comment by yinkev** [comment]
Addressed the latest integer-dtype review finding in `354e5975`. B-spline range arithmetic now promotes non-floating inputs to float32 before `torch.finfo`, preserving mixed floating prediction/integer target support. The new regression failed on the previous head with the reported `TypeError` and now verifies finite loss and prediction gradients. The exact regression passes; the full affected module passes (24 passed, 1 skipped), along with Black, isort, Ruff, and `git diff --check`.
